### DISCUSSION TITLE
how to solve error LNK2005 occurs when linking ktx.dll

### DISCUSSION THREAD

**Discussion by xingfu2005** [vsg-dev/VulkanSceneGraph#1266] [Q&A] [answered]
the external dependencies  of cesium-native are now built using vcpkg.
when building vsgCs use new cesium-native version,then error LNK2005 occurred in ktx.dll:
 ktxTexture_GetData is already defined in vsgXchanged.lib(texture.c.obj)
ktxTexture_GetDataSize is already defined in vsgXchanged.lib(texture.c.obj)
ktxTexture2_CreateFromMemory is already defined in vsgXchanged.lib(texture2.c.obj)
ktxTexture2_GetNumComponents is already defined in vsgXchanged.lib(texture2.c.obj)
 ktxTexture2_NeedsTranscoding is already defined in vsgXchanged.lib(texture2.c.obj)

**Comment by robertosfield** [comment]
This is new issue so I don't have an off the shelf solution.

Would it be possible to hide the KTX symbols built by the vsgXchange::ktx plugin?  Perhaps CMake has an option for this.

**Reply by xingfu2005** [reply, marked as answer]
Thanks your advice!
The problem has been solved. I modified vsgXchange\src\ktx\build_vars.cmake to use external ktx dependency build:

find_package(Ktx CONFIG REQUIRED)
if(Ktx_FOUND)
    set(SOURCES ${SOURCES} ktx/ktx.cpp)
    set(EXTRA_LIBRARIES ${EXTRA_LIBRARIES} KTX::ktx)
    set(EXTRA_DEFINES ${EXTRA_DEFINES}  NEED_VK_GET_FORMAT_SIZE)
    set(EXTRA_INCLUDES ${EXTRA_INCLUDES} $<BUILD_INTERFACE:${VSGXCHANGE_SOURCE_DIR}/src/ktx/lib>)
    if(NOT BUILD_SHARED_LIBS)
        set(FIND_DEPENDENCY ${FIND_DEPENDENCY} "find_dependency(Ktx)")
    endif()
 else()
    # Compile ktx read library
    set(KTX_SOURCES
        ktx/libktx/ktx.h
    #    ktx/libktx/basis_sgd.h
        ktx/libktx/checkheader.c
        ktx/libktx/dfdutils/createdfd.c
    #    ktx/libktx/dfdutils/colourspaces.c
        ktx/libktx/dfdutils/dfd.h
        ktx/libktx/dfdutils/dfd2vk.inl
        ktx/libktx/dfdutils/interpretdfd.c
        ktx/libktx/dfdutils/printdfd.c
        ktx/libktx/dfdutils/queries.c
        ktx/libktx/dfdutils/vk2dfd.c
    #    ktx/libktx/dfdutils/vulkan/vk_platform.h
    #    ktx/libktx/dfdutils/vulkan/vulkan_core.h
    #    ktx/libktx/etcdec.cxx
    #    ktx/libktx/etcunpack.cxx
        ktx/libktx/filestream.c
        ktx/libktx/filestream.h
    #    ktx/libktx/formatsize.h
    #    ktx/libktx/gl_format.h
    #    ktx/libktx/gl_funcs.c
    #    ktx/libktx/gl_funcs.h
    #    ktx/libktx/glloader.c
        ktx/libktx/hashlist.c
    #    ktx/libktx/info.c
        ktx/libktx/ktxint.h
        ktx/libktx/memstream.c
        ktx/libktx/memstream.h
    #    ktx/libktx/stream.h
    #    ktx/libktx/strings.c
        ktx/libktx/swap.c
        ktx/libktx/texture.c
        ktx/libktx/texture.h
        ktx/libktx/texture1.c
        ktx/libktx/texture1.h
        ktx/libktx/texture2.c
        ktx/libktx/texture2.h
    #    ktx/libktx/uthash.h
    #    ktx/libktx/vk_format.h
    #    ktx/libktx/vkformat_check.c
    #    ktx/libktx/vkformat_enum.h
    #    ktx/libktx/vkformat_str.c
    #    ktx/libktx/ktxvulkan.h
        ktx/libktx/vkloader.c
        ktx/libktx/zstddeclib.c
    )
    source_group(libktx FILES ${KTX_SOURCES})
    set(SOURCES ${SOURCES} ${KTX_SOURCES} ktx/ktx.cpp)
    set(EXTRA_DEFINES ${EXTRA_DEFINES} KHRONOS_STATIC LIBKTX BASISD_SUPPORT_FXT1=0 BASISU_NO_ITERATOR_DEBUG_LEVEL KTX_FEATURE_KTX1 KTX_FEATURE_KTX2)
    set(EXTRA_INCLUDES ${EXTRA_INCLUDES} $<BUILD_INTERFACE:${VSGXCHANGE_SOURCE_DIR}/src/ktx/libktx>)
endif()

**Comment by robertosfield** [comment]
Could you generate at Pull Request with these changes. I will try things
out on Linux and see how easy it is now to pull ktx as an external
dependency. If it goes smoothly we can make it the new way of managing ktx
support.
